pkg/audit/reverify_test.go

```diff
@@ -0,0 +1,423 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit_test
+
+import (
+	"crypto/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/pkg/audit"
+	"storj.io/storj/pkg/peertls/tlsopts"
+	"storj.io/storj/pkg/pkcrypto"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+)
+
+func TestReverifySuccess(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount: 1, StorageNodeCount: 6, UplinkCount: 1,
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+
+		// This is a bulky test but all it's doing is:
+		// - uploads random data
+		// - uses the cursor to get a stripe
+		// - creates one pending audit for a node holding a piece for that stripe
+		// - the actual share is downloaded to make sure ExpectedShareHash is correct
+		// - calls reverify on that same stripe
+		// - expects one storage node to be marked as a success in the audit report
+
+		err := planet.Satellites[0].Audit.Service.Close()
+		require.NoError(t, err)
+
+		ul := planet.Uplinks[0]
+		testData := make([]byte, 1*memory.MiB)
+		_, err = rand.Read(testData)
+		require.NoError(t, err)
+
+		err = ul.Upload(ctx, planet.Satellites[0], "testbucket", "test/path", testData)
+		require.NoError(t, err)
+
+		metainfo := planet.Satellites[0].Metainfo.Service
+		cursor := audit.NewCursor(metainfo)
+
+		var stripe *audit.Stripe
+		stripe, _, err = cursor.NextStripe(ctx)
+		require.NoError(t, err)
+		require.NotNil(t, stripe)
+
+		orders := planet.Satellites[0].Orders.Service
+		containment := planet.Satellites[0].DB.Containment()
+
+		verifier := audit.NewVerifier(zap.L(),
+			planet.Satellites[0].Transport,
+			planet.Satellites[0].Overlay.Service,
+			containment,
+			orders,
+			planet.Satellites[0].Identity,
+			128*memory.B,
+			5*time.Second)
+		require.NotNil(t, verifier)
+
+		projects, err := planet.Satellites[0].DB.Console().Projects().GetAll(ctx)
+		require.NoError(t, err)
+
+		bucketID := []byte(storj.JoinPaths(projects[0].ID.String(), "testbucket"))
+		shareSize := stripe.Segment.GetRemote().GetRedundancy().GetErasureShareSize()
+
+		pieces := stripe.Segment.GetRemote().GetRemotePieces()
+		rootPieceID := stripe.Segment.GetRemote().RootPieceId
+		limit, err := orders.CreateAuditOrderLimit(ctx, planet.Satellites[0].Identity.PeerIdentity(), bucketID, pieces[0].NodeId, rootPieceID, shareSize)
+		require.NoError(t, err)
+
+		share, err := verifier.GetShare(ctx, limit, stripe.Index, shareSize, int(pieces[0].PieceNum))
```

Comment:
it would be cool if we could get the share directly from the testdata we generated earlier, but this is probably fine if that's too difficult